Extension systems + Relative Extension systems (#342)
plt-amy/1lab: src/Cat/Diagram/Monad/Relative.lagda.md

<!--
```agda
open import 1Lab.Path.Cartesian

open import Cat.Displayed.Base
open import Cat.Prelude

import Cat.Functor.Reasoning
import Cat.Reasoning

open Functor
open _=>_
```
-->

```agda
module Cat.Diagram.Monad.Relative where
```

# Relative extension systems {defines=relative-extension-system}

By definition, [[monads]] must be endofunctors. For most applications, this
is not relevant, and the theory of monads is quite rich. However, this
restriction can become problematic. For simplicity, let $T$ be a monad
on set interpreted as an algebraic theory, where $T(A)$ is the syntax of
$T$ with $A$-many variables, the unit constructs a term from a variable,
and the join performs substitution. If we wish to give a more refined
treatment of variables^[for instance, if a term can only have finitely
many variables], then we immediately run into an issue: $T$ can no

> longer an endofunctor $\Sets \to \Sets$, and restricting $T$ to

Reviewer: ```suggestion
longer be an endofunctor $\Sets \to \Sets$, and restricting $T$ to
```

$\FinSets \to \FinSets$ is also out of the question, as the collection
of syntax trees will not be finite.

If we wish to capture such situations, then we will need to generalize
the notion of monad. The fundamental problem with doing so is that we

> cannot iterate a functor $M : \cJ \to \cC$, so themost natural starting

Reviewer: ```suggestion
cannot iterate a functor $M : \cJ \to \cC$, so the most natural starting
```

Like their non-relative counterparts, we do not require any functoriality
or naturality.

<!--
```agda
module _
  {oj ℓj oc ℓc}
  {J : Precategory oj ℓj}
  {C : Precategory oc ℓc}
  (F : Functor J C)
  where
  private
    module J = Cat.Reasoning J
    module C = Cat.Reasoning C
    module F = Cat.Functor.Reasoning F 
```
-->

```agda
  record Relative-extension-system : Type (oj ⊔ oc ⊔ ℓc) where
    no-eta-equality
    field
      M₀ : J.Ob → C.Ob
      unit : ∀ {x} → C.Hom (F.₀ x) (M₀ x)
      bind : ∀ {x y} → C.Hom (F.₀ x) (M₀ y) → C.Hom (M₀ x) (M₀ y)
```

The equations for a relative extension system are similar to the
non-relative case.

```agda
      bind-unit-id : ∀ {x} → bind (unit {x}) ≡ C.id
      bind-unit-∘ : ∀ {x y} (f : C.Hom (F.₀ x) (M₀ y)) → bind f C.∘ unit ≡ f
      bind-∘
        : ∀ {x y z}
        → (f : C.Hom (F.₀ y) (M₀ z)) (g : C.Hom (F.₀ x) (M₀ y))
        → bind f C.∘ bind g ≡ bind (bind f C.∘ g)
```

The functorial action of $M$ on $f$ can be recovered by extending
$\eta \circ f$.

```agda
    M₁ : ∀ {x y} → J.Hom x y → C.Hom (M₀ x) (M₀ y)
    M₁ f = bind (unit C.∘ F.₁ f)
```

Furthermore, the latter two equations ensure naturality of the unit and

> extension, resp.

Reviewer: ```suggestion
extension, respectively.
```
